cpu/atmega2560/periph/gpio.c

```diff
@@ -29,47 +31,440 @@
 {
-    return -1;
+    uint8_t pin;
+
+    pin = gpio_pin_map[dev];
+
+    switch (dev)
+    {
+#if GPIO_0_EN
+        case GPIO_0:
+            GPIO_0_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_0_EN */
+
+#if GPIO_1_EN
+        case GPIO_1:
+            GPIO_1_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_1_EN */
+
+#if GPIO_2_EN
+        case GPIO_2:
+            GPIO_2_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_2_EN */
+
+#if GPIO_3_EN
+        case GPIO_3:
+            GPIO_3_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_3_EN */
+
+#if GPIO_4_EN
+        case GPIO_4:
+            GPIO_4_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_4_EN */
+
+#if GPIO_5_EN
+        case GPIO_5:
+            GPIO_5_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_5_EN */
+    };
+
+    return 0;
 }
 
 int gpio_init_in(gpio_t dev, gpio_pp_t pushpull)
 {
-    return -1;
+    uint8_t pin;
+
+    pin = gpio_pin_map[dev];
+
+    switch (dev)
+    {
+#if GPIO_0_EN
+        case GPIO_0:
+            GPIO_0_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_0_EN */
+
+#if GPIO_1_EN
+        case GPIO_1:
+            GPIO_1_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_1_EN */
+
+#if GPIO_2_EN
+        case GPIO_2:
+            GPIO_2_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_2_EN */
+
+#if GPIO_3_EN
+        case GPIO_3:
+            GPIO_3_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_3_EN */
+
+#if GPIO_4_EN
+        case GPIO_4:
+            GPIO_4_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_4_EN */
+
+#if GPIO_5_EN
+        case GPIO_5:
+            GPIO_5_DDR_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_5_EN */
+    };
+
+    return 0;
 }
 
 int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t cb, void *arg)
 {
-    return -1;
+    int ret_code;
+    uint8_t pin_B;
+    uint8_t pin;
+
+    ret_code = gpio_init_in(dev, pushpull);
+    if (ret_code < 0)
+    {
+        return ret_code;
+    }
+
+    pin = gpio_pin_map[dev];
+
+    /* clear global interrupt flag */
+    cli();
+
+    switch(dev)
+    {
+#if GPIO_0_EN
+        case GPIO_0:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_0_EN */
+
+#if GPIO_1_EN
+        case GPIO_1:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_1_EN */
+
+#if GPIO_2_EN
+        case GPIO_2:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_2_EN */
+
+#if GPIO_3_EN
+        case GPIO_3:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_3_EN */
+
+#if GPIO_4_EN
+        case GPIO_4:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_4_EN */
+
+#if GPIO_5_EN
+        case GPIO_5:
+            EIMSK |= (1 << pin);
+            break;
+#endif /* GPIO_5_EN */
+    };
+
+    pin_B = (pin > GPIO_3 && pin == GPIO_4) ? 0 : 1;
```

Comment:
I don't now exactly what you are doing here. Are you sure the `&&` is correct? Anyway, do you know [modulo operations](http://en.wikipedia.org/wiki/Modulo_operation)? Let's do it like this, without using the above line :
```
    /* configure the flank */
    switch (flank) {
        case GPIO_RISING:
            if (pin < 4)
            {
                EICRA |= (3 << pin * 2);
            }
            else
            {
                EICRB |= (3 << (pin * 2)%4));
            }
            break;
```